### PR TITLE
Named arguments support in IDE

### LINKED ISSUE #5083: Add support of Named arguments to Span Tree Generator
This task is automatically imported from the old Task Issue Board and it was originally created by Greg Ociepka.
Original issue is [here](https://www.pivotaltracker.com/story/show/184335109).
<hr />

Add support for `on_problems=value` syntax in node arguments on IDE side. Currently the arguments are always treated as positional, ignoring the name part. This likely depends on new parser being integrated to IDE.

Node that the engine respects the named arguments, so the expressions are evaluated correctly. The issue is purely about how IDE interprets the code present on the node. The visible consequence is invalid argument placeholder and widget assignment.

## Comments:
 (Paweł Grabarz - Jan 27, 2023)
<hr />

### LINKED ISSUE #5098: Unused arguments vanish and replaced with `_` once dropdowns value entered - function then invalid, and parameters vanish
This task is automatically imported from the old Task Issue Board and it was originally created by Hubert Plociniczak.
Original issue is [here](https://www.pivotaltracker.com/story/show/184213021).
<hr />

Consider a situation with simple dropdowns. Selecting a dropdown works, thanks to work on #184023445. But removing the selection has some issues. Namely:
- the automatic insertion of `_` even if only one default argument is deselected
- it's impossible to come back to the pristine view of the node with parameter names

## Comments:
 (Hubert Plociniczak - Jan 12, 2023)
<hr />
 (Hubert Plociniczak - Jan 12, 2023)
<hr />

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

Added support for named arguments in IDE.

https://user-images.githubusercontent.com/919491/223681303-4c716639-d06e-4e33-aa22-6ebca2801b01.mp4

Named arguments are now recognized in node expressions. The function argument placeholders are rendered around series of named arguments. Insertion and deletion of arguments either by connection dragging or by widget selection will cause arguments around to be rewritten into appropriate form, such that the meaning of the expression doesn't change. We no longer need to introduce any wildcards (`_`) in argument positions when editing an argument list of a resolved method.
image

For unresolved function calls, the old behaviour remains, as we don't have data about argument names or their desired order.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: app/gui/language/span-tree/src/generate.rs [hunk 1]
```diff
@@ -448,63 +528,230 @@ fn generate_node_for_prefix_chain<T: Payload>(
...
+                    // we can remove them only if removal preserves the prefix application.
+                    let resolved = info.is_some();
+                    let removable = resolved || !is_only_argument;
+
+                    let mut arg_kind = node::Kind::from(
+                        node::Kind::argument()
+                            .with_removable(removable)
+                            .with_name(arg_name.map(ToString::to_string)),
+                    );
+
```
Comment (wdanilo) [inline app/gui/language/span-tree/src/generate.rs:584]: this is pretty long fn, are we able to split it into several Fns? No worries if not

**Review by wdanilo — APPROVED** [review]
The code is easy to read, has beautiful comments with explanations and I do not have any comments to the code really. Needs QA, but beside that, it's very nice!

File: app/gui/language/span-tree/src/generate.rs (hunk 1 above)
Comment (Frizi) [reply app/gui/language/span-tree/src/generate.rs:584]: I tried, but it only led to scattering of code that has a single place and function. I think it is easier to follow as a single contiugous block.

Comment (wdanilo) [reply app/gui/language/span-tree/src/generate.rs:584]: ok!

**Comment by vitvakatu** [comment]
## QA report

I noticed two issues that may or may not be regressions.

### Colors of edges seem to be random

Notice the color of the `this` argument changes when I connect one of the inputs.

https://user-images.githubusercontent.com/6566674/223846808-d3779132-735a-4356-87e8-8e8ede2e8e38.mp4


### Clicking on the output port creates an edge

I connected the first edge, then I *clicked* on another node's output port – and you can see what happened on the video. After that, clicking on output ports and dragging edges produces weird results without much pattern. Sometimes I even couldn't drop the connection without connecting it to anything. It happens even after reopening the project.


https://user-images.githubusercontent.com/6566674/223846306-99aa55c8-06de-4ae7-aecf-028d6c34878f.mp4


### Config

```
Frontend:
    Version:  2023.1.1-dev
    Build:    333e5a7657a200d5b75956c7098908f2022228cd
    Electron: 23.0.0
    Chrome:   110.0.5481.77

Backend:
    Enso Project Manager
    Version:    2023.1.1-nightly.2023.3.4
    Built with: scala-2.13.8 for GraalVM 22.3.1
    Built from: develop @ f64edd806d4de7e71e6b1fea41a13ac189ebfadd
    Built on:   Mac OS X (amd64)
```

**Comment by wdanilo** [comment]
@vitvakatu amazing job with finding the issues!

**Comment by Frizi** [comment]
@vitvakatu I have reproduced both of those issues on current develop. Those are not introduced in this PR.

https://user-images.githubusercontent.com/919491/224014964-827675e8-df72-4eaa-b7b5-3c8c26fd5fc1.mp4


The reproduction indeed seems a bit random. The edge coloring bug is fairly easy to trigger. The one when clicks cause edges to be modified or added required some more fiddling. I randomly connected and moved things around. After a while, clicking on ports started causing weird behaviours.

**Review by vitvakatu — APPROVED** [review]
QA accepted. The spotted regressions will be reported as issues by me.

**Pull request merged**